### ISSUE TITLE
`ttnn.gelu` accuracy should be improved

### ISSUE THREAD

**Issue by sgligorijevicTT** [tenstorrent/tt-metal#25326]
### Describe the bug

`ttnn.gelu` seems to have accuracy problems in the range [-5, -2].
This seems to be a known problem( #21313  #23669 https://github.com/tenstorrent/tt-metal/issues/20562 ) but I want to have a separate tracking issue on Forge side.
This is a hard blocker for 3 variants of ViT (base patch32 384, large patch16 224 and large patch32 384) and Roberta trough tt-xla. It's interesting that some variants of ViT work fine with current levels of inaccuracy while others don't.
In apparent contradiction to investigations done in #21313 , when we use Forge tooling to bypass gelu and do it on the CPU, ViT passes.

### Steps to reproduce the issue

1. Run the following code
```
import torch
import ttnn
import numpy as np

def spacing(x: torch.Tensor):
    """Calculate spacing (ULP size) for bfloat16 and float32 inputs."""
    # If input is bfloat16, use abs(x) * 2^-7 for normal numbers
    if x.dtype == torch.bfloat16:
        ulp = torch.abs(x).to(torch.float32) * 2**-7
        # For zero, use smallest positive normal bfloat16 (2^-126)
        ulp = torch.where(x == 0, torch.tensor(2**-126, dtype=torch.float32), ulp)
        return ulp
    else:
        # For float32, use nextafter
        direction = torch.where(x >= 0, torch.inf, -torch.inf)
        next_val = torch.nextafter(x, direction)
        return torch.abs(next_val - x)

def ulp_difference(a, b):
    """Calculate ULP (Units in the Last Place) difference between two tensors for bfloat16 and float32."""
    # Ensure inputs are torch tensors
    if not torch.is_tensor(a):
        a = torch.tensor(a)
    if not torch.is_tensor(b):
        b = torch.tensor(b)
    # Move to CPU and ensure same device/dtype
    a = a.detach().cpu()
    b = b.detach().cpu()
    # Handle special cases (inf, nan)
    mask_valid = torch.isfinite(a) & torch.isfinite(b)
    if not torch.any(mask_valid):
        return torch.tensor([float('inf')])
    # Use the larger absolute value for spacing calculation
    max_abs = torch.maximum(torch.abs(a), torch.abs(b))
    # Calculate spacing (distance to next representable float)
    ulp_spacing = spacing(max_abs)
    # Calculate ULP difference
    abs_diff = torch.abs(a - b)
    ulp_diff = abs_diff / ulp_spacing
    return ulp_diff[mask_valid]

RANGE = 4

def test_ttnn_gelu():
    x_values = torch.linspace(-RANGE, RANGE, 2048, dtype=torch.bfloat16).reshape(-1, 32, 32)
    
    with ttnn.manage_device(0) as dev:
        # Convert to ttnn tensor with BFLOAT16 and tile layout
        ttnn_input = ttnn.from_torch(
            x_values,
            dtype=ttnn.bfloat16,
            layout=ttnn.TILE_LAYOUT,
            device=dev
        )
    
        ttnn_result = ttnn.gelu(ttnn_input)
        ttnn_output = ttnn.to_torch(ttnn_result)
        torch_reference = torch.nn.functional.gelu(x_values)
        
        print("\n" + "="*60)
        print("COMPARISON RESULTS")
        print("="*60)
        
        # Calculate various error metrics
        abs_diff = torch.abs(ttnn_output - torch_reference)
        rel_diff = abs_diff.float() / (torch.abs(torch_reference.float()) + 1e-9)
        
        print(f"Max absolute difference: {abs_diff.max():.6e}")
        print(f"Mean absolute difference: {abs_diff.mean():.6e}")
        print(f"Max relative difference: {rel_diff.max():.6e}")
        print(f"Mean relative difference: {rel_diff.mean():.6e}")
        
        # Calculate ULP difference
        ulp_diffs = ulp_difference(ttnn_output, torch_reference)
        
        print(f"\nULP Statistics:")
        print(f"Max ULP difference: {torch.max(ulp_diffs):.2f}")
        print(f"Mean ULP difference: {torch.mean(ulp_diffs):.2f}")
        print(f"Median ULP difference: {torch.median(ulp_diffs):.2f}")
        
        N=30
        # Print some sample comparisons (equally spaced subset)
        total_values = len(x_values.flatten())
        indices = np.linspace(0, total_values - 1, N, dtype=int)
        
        print(f"\nSample comparisons ({N} equally spaced values):")
        print("Input\t\tTorch gelu\tTTNN gelu\tAbs Diff\tULP Diff\tOne ULP size")
        print("-" * 70)
        for i in indices:
            inp = x_values.flatten()[i].item()
            torch_val = torch_reference.flatten()[i].item()
            ttnn_val = ttnn_output.flatten()[i].item()
            abs_d = abs_diff.flatten()[i].item()
            ulp_d = ulp_diffs[i] if i < len(ulp_diffs) else 0
            torch_ulp = spacing(torch.tensor(torch_val, dtype=torch.bfloat16)).item()
            print(f"{inp:8.3f}\t{torch_val:10.6f}\t{ttnn_val:10.6f}\t{abs_d:.2e}\t{ulp_d:8.2f}\t{torch_ulp:.2e}")
        
        reasonable_ulp = torch.sum(ulp_diffs < 50) / len(ulp_diffs) * 100
        print(f"\nPercentage of values with ULP < 50: {reasonable_ulp:.1f}%")


if __name__ == "__main__":
    test_ttnn_gelu()
```

### Expected behavior

I don't know what level of accuracy I need, but I need it improved. Attached code is for illustration purposes only.

### Please complete the following environment information

Ubuntu 22.04
Metal hash d5c259dd7bc8d0c9879a4389d50e65c321cc0c6f

**Comment by sgligorijevicTT** [comment]
As ViTs are RED, I'm marking as P1.

**Comment by mcw-vishal** [comment]
Update:- We are working on improving the performance of chebychev poly15 gelu approach mentioned in issue https://github.com/tenstorrent/tt-metal/issues/20562. Currently it is 41% slower. I have compared the experiment with chebychev poly15 approach.

| ULP Threshold         | Default Gelu (%) | ChebyChev Polyval15 Gelu (%) |
|-----------------------|------------------|---------------------|
| ULP < 50              | 81.1             | 99.5                |
| ULP < 10              | 80.3             | 92.8                |
| ULP < 5               | 78.0             | 90.5                |
| ULP < 2               | 74.8             | 89.0                |

**Comment by mcw-vishal** [comment]
Code changes are in the [branch](https://github.com/tenstorrent/tt-metal/tree/vishal/gelu-chebychev)

**Comment by mcw-vishal** [comment]
Update:- Compared default ttnn.gelu with chebychev polyval15 gelu on various ranges and metrics:-

<img width="3564" height="2364" alt="Image" src="https://github.com/user-attachments/assets/a6144469-91eb-4511-8762-807ff4194aa1" />
<img width="3564" height="2364" alt="Image" src="https://github.com/user-attachments/assets/2f06fba7-e37e-4aed-9b71-98371034cef5" />
<img width="3564" height="2364" alt="Image" src="https://github.com/user-attachments/assets/76ef0e7e-f4df-43e8-8a61-7b948803ea87" />
<img width="3564" height="2364" alt="Image" src="https://github.com/user-attachments/assets/3bc953e8-2abe-4521-926d-ecd802ccc7ab" />
<img width="3566" height="2364" alt="Image" src="https://github.com/user-attachments/assets/9a4f4251-7477-49f9-a8d6-82e127eba3cc" />
<img width="3562" height="2364" alt="Image" src="https://github.com/user-attachments/assets/0a238449-5ac3-48bb-87bd-de62b45c75c2" />
<img width="3564" height="2364" alt="Image" src="https://github.com/user-attachments/assets/058ca5c0-7ffa-462b-b73e-f258ba841e14" />
<img width="3564" height="2364" alt="Image" src="https://github.com/user-attachments/assets/0a33072c-38bd-4bc1-b403-6733c1b6dbda" />
<img width="3564" height="2364" alt="Image" src="https://github.com/user-attachments/assets/bc8f40cd-7031-47aa-bdc0-6c9debdcdf3b" />

**Comment by mcw-vishal** [comment]
| Metric                         | Default Gelu       | Chebychev polyval15 Gelu   |
|-------------------------------|--------------------|--------------------|
| Max absolute error            | 0.01562500         | 0.01562500         |
| Mean absolute error           | 0.00110746         | 0.00072949         |
| Standard deviation            | 0.00326050         | 0.00272597         |
| 95th percentile               | 0.00781250         | 0.00390625         |
| 99th percentile               | 0.01562500         | 0.01562500         |
| Points > 0.010 (count/%)      | 8 (3.98%)          | 5 (2.49%)          |
| Points > 0.020 (count/%)      | 0 (0.00%)          | 0 (0.00%)          |
| Points > 0.030 (count/%)      | 0 (0.00%)          | 0 (0.00%)          |
| Points > 0.040 (count/%)      | 0 (0.00%)          | 0 (0.00%)          |
| Points > 0.050 (count/%)      | 0 (0.00%)          | 0 (0.00%)          |

| Comparison Metric                        | Count | Percentage |
|------------------------------------------|-------|------------|
| Default Gelu is better                   | 20    | 10.0%      |
| Chebychev polyval15 Gelu is better               | 33    | 16.4%      |
| Equal performance                        | 148   | 73.6%      |

**Comment by mcw-vishal** [comment]
ULP plots for default and chebychev polyval15 gelu

<img width="1000" height="500" alt="Image" src="https://github.com/user-attachments/assets/891541db-1a59-4863-94d1-7ebb04f982b8" />

<img width="1000" height="500" alt="Image" src="https://github.com/user-attachments/assets/6fa4ffe9-299d-4d15-9d04-4c30a808fe0f" />

**Comment by fahmedchanTT** [comment]
@cmaryanTT or @mcw-vishal What are the next steps here for this case?

**Comment by fahmedchanTT** [comment]
@mcw-vishal Just following up here.

**Comment by cmaryanTT** [comment]
@fahmedchanTT - on hold, need to re-evaluate from first principles to see if we can keep perf reasonable while cleaning up accuracy.

**Comment by mcw-vishal** [comment]
# Chebyshev Implementation Performance Analysis

## Overview
Performance comparison of new Chebyshev approximation approaches (BH, WH) against baseline implementations.

## Baseline vs Chebyshev Approaches

| Implementation | 1 Tile | 1K Tensor | 8K Tensor | Type |
|----------------|--------|-----------|-----------|------|
| **Main Branch (WH)** | 4,465 ns | 52,048 ns | 3,054,703 ns | 🔵 Baseline |
| **Main Branch Approx (WH)** | 1,866 ns | 26,849 ns | 1,340,607 ns | 🔵 Baseline |
| **Chebyshev BH** | 2,654 ns | 21,584 ns | 844,148 ns | 🟢 New Approach |
| **Chebyshev WH** | 3,841 ns | 41,749 ns | 2,373,080 ns | 🟢 New Approach |

## Chebyshev Performance Gains

### vs Main Branch (WH) Baseline
| Chebyshev Approach | 1 Tile | 1K Tensor | 8K Tensor |
|-------------------|--------|-----------|-----------|
| **BH Implementation** | 🟢 **1.68x faster** | 🟢 **2.41x faster** | 🟢 **3.62x faster** |
| **WH Implementation** | 🟢 **1.16x faster** | 🟢 **1.25x faster** | 🟢 **1.29x faster** |

### vs Main Branch Approx (WH) Baseline
| Chebyshev Approach | 1 Tile | 1K Tensor | 8K Tensor |
|-------------------|--------|-----------|-----------|
| **BH Implementation** | 🔴 **1.42x slower** | 🟢 **1.24x faster** | 🟢 **1.59x faster** |
| **WH Implementation** | 🔴 **2.06x slower** | 🔴 **1.56x slower** | 🔴 **1.77x slower** |

**Comment by fahmedchanTT** [comment]
Thanks @mcw-vishal for performance numbers.
What is the next step here?

**Comment by mcw-vishal** [comment]
Hi @fahmedchanTT I have created a draft PR for the chebychev accurate implementation of gelu. The integrating this approach with models is tricky part. I am checking with the model owners (llama3, segformer, qwen25vl and others) regarding the same.

Please find the draft pr [link](https://github.com/tenstorrent/tt-metal/pull/27283)

**Comment by mcw-vishal** [comment]
Update:-

- **Qwen25_vl tests**
  - The model with chebychev gelu generated better text output.  Confirmed with respective model owners we can update the expectation in the test (Attaching screenshot, text in red is older).

<img width="1447" height="295" alt="Image" src="https://github.com/user-attachments/assets/091a8367-64e0-4383-9bde-99c9327ebc59" />

- **Segformer model tests** 
  - Expected pcc is 0.978 and computed pcc is 0.977859290229
  - During investigation it was found that the model is dependent on zero outputs ( earlier gelu was squashing input <-2.5 to 0) and now we have increased the range till -5.5. 

- **Llama3 tests** 
  - Llama3_90b on main branch is already having the regression, not dependent on gelu changes. (Confirmed with model owner)
  - Llama3_vision is already having regression on Main Branch
    - t3000-demo-tests / t3k_llama3_vision_tests`decode_t/s (8.33677127495482) is lower than expected 34.25`
2025-09-04T12:11:15.9049230Z 2025-09-04 12:11:15.901 | WARNING  | models.demos.utils.llm_demo_utils:verify_perf:179 - `decode_t/s/u (8.33677127495482) is lower than expected 34.25`
  - On Gelu chebychev regression is similar
decode_t/s (8.33677127495482) is lower than expected 34.25
2025-09-04T12:05:09.8877379Z 2025-09-04 12:05:09.884 | WARNING  | models.demos.utils.llm_demo_utils:verify_perf:179 - `decode_t/s (5.74981237989852) is lower than expected 34.25`

**Comment by mcw-vishal** [comment]
Update:- Attaching the ULP graphs for GELU (Accurate Main) and Chebychev GELU.
Following points can be observed:-

- There are no ULP Error recorded for values <-2.5 in Main branch gelu, the reason is the squashing at -2.5f.
-  Main Branch has significant ULP errors at inputs near -2.0 range and ChebyChev GELU has lower ULP errors.
- However, there are some ULP outliers in the range [-4, -3].

<img width="3600" height="1800" alt="Image" src="https://github.com/user-attachments/assets/bcceb080-1fce-465e-9417-0459d21ea330" />

<img width="3600" height="1800" alt="Image" src="https://github.com/user-attachments/assets/0b738503-3dc7-4332-919e-aa6e9aab68db" />

**Comment by mcw-vishal** [comment]
Currently working to improve the ULP for the range [-4, -3] for ChebyChev GELU.

**Comment by mcw-vishal** [comment]
# GELU Implementation Error Metric Analysis

## Implementations Analyzed

1. **TTNN MAIN (Accurate GELU)** - The current main implementation
2. **Chebychev Accurate GELU** - Alternative implementation using Chebyshev approximation

## Metrics

- **MAE**: Mean Absolute Error
- **MSE**: Mean Squared Error  
- **RMSE**: Root Mean Squared Error

## Results

### TTNN MAIN (Accurate GELU)

| Range | Number of Samples | MAE | MSE | RMSE |
|-------|------------------|-----|-----|------|
| [-5.5, -2.5] | 144 | 0.0021921138 | 1.4626324e-05 | 0.0038244377 |
| [-2.5, 0] | 16,416 | 1.819501e-05 | 1.4966226e-08 | 0.00012233653 |
| [0, 3] | 16,449 | 7.853978e-05 | 4.263635e-07 | 0.00065296516 |
| [3, ∞] | 16,191 | 5.0008707e-06 | 1.0938491e-08 | 0.00010458724 |

### Chebychev Accurate GELU

| Range | Number of Samples | MAE | MSE | RMSE |
|-------|------------------|-----|-----|------|
| [-5.5, -2.5] | 144 | 6.995659e-05 | 6.802656e-09 | 8.247822e-05 |
| [-2.5, 0] | 16,416 | 4.4754463e-05 | 9.11358e-09 | 9.5465075e-05 |
| [0, 3] | 16,449 | 0.00011290787 | 4.8382356e-07 | 0.0006955743 |
| [3, ∞] | 16,191 | 5.0008707e-06 | 1.0938491e-08 | 0.00010458724 |

## Key Observations

### Performance Comparison

- **Negative ranges ([-5.5, -2.5])**: Chebychev implementation shows significantly better accuracy
  - MAE improvement: ~31x better (6.995659e-05 vs 0.0021921138)
  - MSE improvement: ~2149x better (6.802656e-09 vs 1.4626324e-05)
  
- **Negative to zero range ([-2.5, 0])**: Chebychev shows modest improvement
  - MAE improvement: ~2.5x better (4.4754463e-05 vs 1.819501e-05)
  
- **Positive range ([0, 3])**: Both implementations show similar performance
  - MAE: Comparable accuracy levels
  - MSE: Similar error magnitudes
  
- **Large positive values ([3, ∞])**: Identical performance
  - Both implementations achieve the same precision for large positive inputs

### Conclusion

The Chebychev Accurate GELU implementation demonstrates superior accuracy, particularly in negative input ranges, while maintaining comparable performance in positive ranges. The most significant improvement is observed in the strongly negative range [-5.5, -2.5], where the Chebychev method achieves substantially lower error rates.

**Comment by fahmedchanTT** [comment]
@mcw-vishal Is this ready to close? 
Is there a PR up for using Chebychev for all instances?

I see your analysis is very thorough, thank you.

**Comment by mcw-vishal** [comment]
> [@mcw-vishal](https://github.com/mcw-vishal) Is this ready to close? Is there a PR up for using Chebychev for all instances?
> 
> I see your analysis is very thorough, thank you.

Hi @fahmedchanTT the chebychev approach is to replace the default/accurate gelu (not the approximate version) of the op. Currently there are some CI related things that are ongoing.
